### ISSUE TITLE
The function ft_channelrepair requires a layout

### ISSUE THREAD

**Issue by m-topor** [fieldtrip/fieldtrip#1774]
**Describe the bug**
I have been trying to overcome an error when using the `ft_channelrepair` funtion for some time now and I think that there might be a bug. I am analysing data from a 32 channel Mobita EEG system.

The documentation for the function suggests that the only two configurations that need to be specified are `cfg.badchannel` and `cfg.neighbour` and there are default options for all other configurations. 

However, I cannot run interpolation without specifying the channel layout. It is also worth noting that there are no channels missing here so the function should interpolate and not reconstruct channels that are not there. The error suggests that the 3D electrode structure should be used:

``cfg = [];``
``cfg.neighbours      = neighbours_EEG;``
``cfg.badchannel      = badchannels;``
``mydata_interpolated_train = ft_channelrepair(cfg, mydata_artrm_train);``

``Error using ft_fetch_sens (line 202)``
``no electrodes, gradiometers or optodes specified.``

I understand that this should not be required? If I use the 2D layout file with `cfg.layout` the function works fine. I can use the "weighted" and "average" methods with the 2D layout configuration, but I cannot use the "spline" method. I wrote the script originally around 6 month ago using the April 2020 version of Fieldtrip and I was using the "spline" method originally without a 3D layout and it seemed to have worked ok. I am continuing my analyses now (though upgraded to a newer Fieldtrip version) and would like to keep the methodology consistent if possible.  

_Also for the record, I asked about this issue on the mailing list last week and there was another person who had a similar issue at the same time - they received the same error as the one above thus asking to provide the elec sturcture file for interpolating an EEG channel using `ft_channelreapir`- which should not have been needed._

**Environment and versions:**
 - Operating System: Windows 10
 - MATLAB version: R2020a
 - FieldTrip version: 20210525

**Comment by robertoostenveld** [comment]
can you post the full error message, including the so-called stack trace that shows how all functions call each other? 

Or can you confirm that this is due to the code at https://github.com/fieldtrip/fieldtrip/blob/831cf37e8ebca35d70e2afd74710b729e07df62f/ft_channelrepair.m#L134 ? 

It indeed seems that the checks on the data or cfg having sensor information are too strict (already since prior to commit 3e710ecc7d in 2016), since the sensors (or layout) are only needed to pass them to ft_prepare_neighbours, which should only be called in case there are no neighbors specified

**Comment by schoffelen** [comment]
just for reference, I am pasting in the e-mails that have been sent back and forth on the discussion list:

Hi Jan, 

Thank you so much for your insightful response. It all makes much more sense now - I apologise for my confusion with this.

My preference for sticking with the "spline" method was motivated by the fact that I had already processed over 40 files using the initially provided configurations and I wanted to keep my methodology consistent going forward with the project. 

I can confirm that changing the method to average and replacing the cfg.elec configuration with cfg.layout ​works fine - which is great news! 

Thanks again for your help and all the best!

_________________________
Marta Topor
PhD Candidate

Room 07AC04
School of Psychology
University of Surrey

**Comment by m-topor** [comment]
> can you post the full error message, including the so-called stack trace that shows how all functions call each other?
> 
> Or can you confirm that this is due to the code at
> 
> https://github.com/fieldtrip/fieldtrip/blob/831cf37e8ebca35d70e2afd74710b729e07df62f/ft_channelrepair.m#L134
> 
> ?
> It indeed seems that the checks on the data or cfg having sensor information are too strict (already since prior to commit [3e710ec](https://github.com/fieldtrip/fieldtrip/commit/3e710ecc7db59905110527bf73cf87978d8a4fb5) in 2016), since the sensors (or layout) are only needed to pass them to ft_prepare_neighbours, which should only be called in case there are no neighbors specified

Yes that's the one @robertoostenveld 
The full error message is as follows:
``Error using ft_fetch_sens (line 202)
no electrodes, gradiometers or optodes specified.``

``Error in ft_channelrepair (line 138)
    sens = ft_fetch_sens(cfg, data);``

**Comment by gemamijancos** [comment]
Hi, 

I also had a similar problem. This is the e-mail I sent:

Currently, we are analysing TMS-EEG signals.

I have a bad channel that I would like to repair, using ft_channelrepair.
This is the code that I used:

cfg = [];
cfg.channel = 'all'
cfg.method = 'distance';
cfg.layout = 'elec1010';
cfg.neighbourdist = 10;
neighbours = ft_prepare_neighbours(cfg, data_tms_raw_sim);


cfg = [];
cfg.method = 'weighted';
cfg.badchannel = 'C1';
cfg.neighbours = neighbours;
cfg.trials = 'all';
interp = ft_channelrepair(cfg, data_tms_raw_sim);

As I do have a channel (is not a missing one) I did not include cfg.elec into the cfg (because reading the documentation it does not say is compulsory to add it). However, I get the following error, which I think is related to cfg.elec. I do not understand why I get this error:

Error in ft_channelrepair (line 136)

    sens = ft_fetch_sens(cfg, data);

Nevertheless, I decided to include cfg.elec:

cfg.elec = 'C:\Users\MIJANCOS MARTINEZ 2\Desktop\BECA\MATLAB CODE\Scripts\Channel_info_EEG_64_sens');

which is a Matlab structure containing the label of the channel, their position and the position of the electrodes (information needed by ft_datatype_sens: sens.lablel, sens.chanpos and sens.elecpos).


However, when I run this code, I get another error:

Error using cell/unique>celluniqueR2012a (line 249)

Elements of cell array input A must be character vectors (row vectors of class char).

Error in cell/unique (line 142)

        [varargout{1:nlhs}] = celluniqueR2012a(varargin{1},logical(flaginds(1:5)));

Error in cell/setdiff>cellsetdiffR2012a (line 303)

    uA = unique(a,order);

Error in cell/setdiff (line 129)

        [varargout{1:nlhs}] = cellsetdiffR2012a(varargin{1:2},logical(flaginds(1:3)));

Error in ft_channelrepair (line 162)

cfg.missingchannel = setdiff(cfg.missingchannel, data.label, 'stable');

Error in Preprocessing_TMS_EEG (line 137)

    [interp] = ft_channelrepair(cfg, data_tms_raw_sim);

I have read some threads regarding this issue, and one of the solutions was to use cfg.layout instead of cfg.elec. Thus, I used the easycap-M10.txt available in Fieldtrip/template but I get the same error as before.
I would really appreciate if anyone could provide some help in understanding why I need to provide cfg.elec (if this is what I need to do, perhaps I am wrong in my initial assumption) and why I am getting the errors that I just mentioned.

The response I got: 

I think that using the non-preferred ‘weighted’ method in ft_channelrepair does not require an elec description. However, you should specify cfg.badchannel as a cell-array (at least that’s what I gather from the function’s help, and what I think I understand from the low-level errors you report).
In other words, rather than cfg.badchannel = ‘C1’, you should call cfg.badchannel = {‘C1’};

Then I would think that the following could work...
cfg = [];
cfg.method = 'weighted';
cfg.badchannel = {‘C1'};
cfg.neighbours = neighbours;
cfg.trials = 'all';
interp = ft_channelrepair(cfg, data_tms_raw_sim);

Nevertheless, by doing so, I got the same error as the beginning, which was:
Error in ft_channelrepair (line 136)

    sens = ft_fetch_sens(cfg, data);

**Comment by schoffelen** [comment]
And this is the e-mail I answered:

Hi Gema,

It seems that ft_channelrepair is in the air today. Just for the record, this will be the last e-mail I will answer on the discussion list today, specifically one about ft_channelrepair…

I think that using the non-preferred ‘weighted’ method in ft_channelrepair does not require an elec description. However, you should specifiy cfg.badchannel as a cell-array (at least that’s what I gather from the function’s help, and what I think I understand from the low level errors you report).
In other words, rather than cfg.badchannel = ‘C1’, you should call cfg.badchannel = {‘C1’};

Then I would think that the following could work...
cfg = [];
cfg.method = 'weighted';
cfg.badchannel = {‘C1'};
cfg.neighbours = neighbours;
cfg.trials = 'all';
interp = ft_channelrepair(cfg, data_tms_raw_sim);


Best wishes,
Jan-Mathijs

**Comment by m-topor** [comment]
Just an update @robertoostenveld - I have evaluated the data I preprocessed previously with my earlier code and I realised that I must have use a 3D elec structure for the "spline" method - which is why I think it initially worked. But I don't think that the file I used was correct for my layout so I will re-run preprocessing with the "weighted" method instead and apply the 2D layout file as currently required.

**Comment by schoffelen** [comment]
OK, as an update for @m-topor @gemamijancos : I just created a PR, which 
1. has an improved docstring in ft_channelrepair, explaining in which cases a cfg.elec/data.elec (or grad/opto) is needed
2. has the code reorganized a bit internally, to be more error prone.

Now, there's one remaining oddity with respect to this issue report @m-topor mentions that it used to be able to use a cfg.layout for this function to work (I guess as a proxy for cfg.elec). I can confirm that this seems to be the case, i.e. the current code in the PR manages to do a 'weighted' interpolation using data without a grad, yet a cfg with a layout only. Yet, this strikes me as odd, because a 2D layout is a very poor proxy for the real distances between electrodes/sensors, thus will probably lead to 'distorted' results. Also, it is nowhere mentioned in the documentation, so I don't understand where @m-topor got this from to begin with. 
So the remaining question now is: should we discontinue the implicit support for using a layout as a proxy for a channel array description? @robertoostenveld, what do you think?

**Comment by robertoostenveld** [comment]
This morning I fixed another regression error https://github.com/fieldtrip/fieldtrip/commit/a11851d4d67f2d59229b357feb7497149320f91e. An old test script that worked fine before started failing (because MEG and EEG being confusing). It should not all work, and hopefully the updated version will test OK and be released tonight.
